mips/signal.c

```diff
@@ -0,0 +1,86 @@
+#include <signal.h>
+#include <thread.h>
+#include <systm.h>
+#include <stdc.h>
+#include <errno.h>
+
+#define SIG_CTX_MAGIC 0xDACBAEE3
+
+typedef struct sig_ctx {
+  uint32_t magic; /* Integrity control. */
+  exc_frame_t ctx;
+  fpu_ctx_t ctx_fpu;
+  /* TODO: Store signal mask. */
+  /* TODO: Store previous stack data, if the sigaction requested a different
+   * stack. */
+} sig_ctx_t;
+
+/* Delivers a signal to user process. */
+int platform_sendsig(int sig, sigaction_t *sa) {
+  thread_t *td = thread_self();
+  assert(mtx_owned(&td->td_lock));
+
+  /* Prepare signal context. */
+  sig_ctx_t ksc;
+
+  ksc.magic = SIG_CTX_MAGIC;
+  ksc.ctx = td->td_uctx;
+  ksc.ctx_fpu = td->td_uctx_fpu;
+
+  /* Copyout signal context to user stack. */
+  sig_ctx_t *scp = (sig_ctx_t *)td->td_uctx.sp;
+  scp--;
+
+  int error = copyout(&ksc, scp, sizeof(sig_ctx_t));
+  if (error) {
+    /* This thread has a corrupted stack, we can no longer send any signals to
+       it. Forcefully terminate the thread. TODO: Maybe sending a SIGKILL is
+       safe here? */
```

Comment:
IMO it should be safe to deliver `SIGKILL` in this case, as it will be handled before a thread has an opportunity to use corrupted stack.